demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c

```diff
@@ -0,0 +1,285 @@
+    if( ( xResult == eAzureIoTSuccess ) && xIsLastBlock && ( pucNextWriteAddr > pucLastChunkWriteAddr + azureiotflashL475_FLASH_ROW_SIZE ) )
+    {
+        while( pucNextWriteAddr < pucBlockEndAddr + azureiotflashL475_FLASH_ROW_SIZE )
+        {
+            /* Program double words until the end of the file */
+            if( HAL_FLASH_Program( FLASH_TYPEPROGRAM_DOUBLEWORD, ( uint32_t ) pucNextWriteAddr, ( uint64_t ) *( uint32_t * ) pucNextReadAddr | ( ( uint64_t ) *( uint32_t * ) ( pucNextReadAddr + 4 ) ) << 32 ) != HAL_OK )
+            {
+                /* Error occurred while writing data in Flash memory */
+                xResult = eAzureIoTErrorFailed;
+                break;
+            }
+
+            pucNextWriteAddr += azureiotflashL475_DOUBLE_WORD_SIZE;
+            pucNextReadAddr += azureiotflashL475_DOUBLE_WORD_SIZE;
+        }
+    }
+
+    HAL_FLASH_Lock();
+
+    return xResult;
+}
+
+AzureIoTResult_t AzureIoTPlatform_VerifyImage( AzureADUImage_t * const pxAduImage,
+                                               uint8_t * pucSHA256Hash,
+                                               uint32_t ulSHA256HashLength )
+{
+    int xResult;
+    uint32_t ulOutputSize;
+    uint32_t ulReadSize;
+
+    AZLogInfo( ( "Base64 Encoded Hash from ADU: %.*s", ulSHA256HashLength, pucSHA256Hash ) );
+    xResult = prvBase64Decode( pucSHA256Hash, ulSHA256HashLength, ucDecodedManifestHash, azureiotflashSHA_256_SIZE, ( size_t * ) &ulOutputSize );
+
+    if( xResult != eAzureIoTSuccess )
+    {
+        AZLogError( ( "Unable to decode base64 SHA256\r\n" ) );
+        return eAzureIoTErrorFailed;
+    }
+
+    mbedtls_md_context_t ctx;
+    mbedtls_md_type_t md_type = MBEDTLS_MD_SHA256;
+
+    mbedtls_md_init( &ctx );
+    mbedtls_md_setup( &ctx, mbedtls_md_info_from_type( md_type ), 0 );
+    mbedtls_md_starts( &ctx );
+
+    AZLogInfo( ( "Starting the mbedtls calculation: image size %d\r\n", pxAduImage->ulImageFileSize ) );
+
+    for( size_t ulOffset = 0; ulOffset < pxAduImage->ulImageFileSize; ulOffset += sizeof( ucPartitionReadBuffer ) )
+    {
+        ulReadSize = pxAduImage->ulImageFileSize - ulOffset < sizeof( ucPartitionReadBuffer ) ? pxAduImage->ulImageFileSize - ulOffset : sizeof( ucPartitionReadBuffer );
+
+        memcpy( ucPartitionReadBuffer, ( pxAduImage->xUpdatePartition + ulOffset ), ulReadSize );
+
+        mbedtls_md_update( &ctx, ( const unsigned char * ) ucPartitionReadBuffer, ulReadSize );
+
+        if( ( ulOffset % 65536 == 0 ) && ( ulOffset != 0 ) )
+        {
+            printf( "." );
+        }
+    }
+
+    printf( "\r\n" );
+
+    AZLogInfo( ( "Done\r\n" ) );
+
+    mbedtls_md_finish( &ctx, ucCalculatedHash );
+    mbedtls_md_free( &ctx );
+
+    if( memcmp( ucDecodedManifestHash, ucCalculatedHash, azureiotflashSHA_256_SIZE ) == 0 )
+    {
+        AZLogInfo( ( "SHAs match\r\n" ) );
+        xResult = eAzureIoTSuccess;
+    }
+    else
+    {
+        AZLogInfo( ( "SHAs do not match\r\n" ) );
```

Comment:
This should not be Information but error (or at least warning). We should also try to log a single time instead of many log lines.

Also, formatting should only be done by AZLogInfo and not printf - a log listener may direct these messages elsewhere.

Reply:
This was done to match the verification in the ESP32 sample, but let me see what I can do to modify it and then make those changes there as well

Reply:
I was able to change the first line to an error (I wasn't sure if the rest should be categorized more as additional info or more error lines).
I couldn't figure out a way to print the whole hash so we could use one logError for the whole bit of info - I think I'm running into this issue: https://eevblog.com/forum/microcontrollers/printf-uint64_t-with-arm-gcc/ 
> AZLogError( ( "SHAs do not match\r\nWanted: %.*x\r\nCalculated: %.*x\r\n", azureiotflashSHA_256_SIZE, *ucDecodedManifestHash, azureiotflashSHA_256_SIZE, *ucCalculatedHash ) );

I also tried to at least convert the printfs to AzureLogInfos, but that adds a newline at the end, so the hash ends up like this instead of on the same line 
> ea
fa
d1
fa
cd
ea
6a
cc
ab
33
...

If you have any suggestions on how to solve either of these issues, let me know!

Reply:
Ah, in that case you could consider the issue a future enhancement and track separately. 

In general, we should never log more than one line per error/info message. This helps log analysis tools that usually expect a single record per message.